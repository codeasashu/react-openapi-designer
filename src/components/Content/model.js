//@flow
import React from 'react';
import PropTypes from 'prop-types';
import {observer} from 'mobx-react-lite';
import SchemaDesigner from '../Editor/JsonSchema';
import {TitleEditor, MarkdownEditor} from '../Editor';
import {usePatchOperation, getValueFromStore} from '../../utils/selectors';
import {nodeOperations} from '../../datasets/tree';

const ModelContent = observer(({relativeJsonPath, node}) => {
  const handlePatch = usePatchOperation();
  const schema = getValueFromStore(relativeJsonPath);

  return schema ? (
    <div className="flex-1 relative">
      <div className="EditorPanel EditorPanel--primary EditorPanel--forms group p-0 flex flex-col relative inset-0">
        <div className="max-w-6xl p-10 flex flex-col">
          <div className="flex pl-2 justify-between">
            <TitleEditor
              xl
              role="title"
              value={schema?.title || node.path}
              onChange={(e) => {
                handlePatch(
                  nodeOperations.Replace,
                  relativeJsonPath.concat(['title']),
                  e.target.value,
                );
              }}
            />
          </div>
          <div className="flex-1">
            <div className="mt-6">
              <MarkdownEditor
                className="CodeEditor mb-8 relative hover:bg-darken-2 rounded-lg"
                placeholder="Description...."
                value={schema?.description || ''}
                onChange={(e) =>
                  handlePatch(
                    nodeOperations.Replace,
                    relativeJsonPath.concat(['description']),
                    e,
                  )
                }
              />
            </div>
          </div>
          <SchemaDesigner
            schemaPath={relativeJsonPath}
            examplesPath={relativeJsonPath.concat('examples')}
            className="mt-6"
          />
        </div>
      </div>
    </div>
  ) : (
    <p>No Schema found</p>
  );
});

ModelContent.propTypes = {
  relativeJsonPath: PropTypes.array,
  node: PropTypes.object,
};

export default ModelContent;
