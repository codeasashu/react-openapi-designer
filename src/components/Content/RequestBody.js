//@flow
import React from 'react';
import PropTypes from 'prop-types';
import RequestBody from './operation/requestBody';

const RequestBodyContent = ({relativeJsonPath}) => {
  const contentPath = relativeJsonPath.concat(['content']);
  const descriptionPath = relativeJsonPath.concat(['description']);
  return (
    <div className="flex-1 relative">
      <div className="EditorPanel EditorPanel--primary EditorPanel--forms group p-0 flex flex-col relative inset-0">
        <div className="max-w-6xl p-10 flex flex-col">
          <RequestBody
            contentPath={contentPath}
            descriptionPath={descriptionPath}
          />
        </div>
      </div>
    </div>
  );
};

RequestBodyContent.propTypes = {
  relativeJsonPath: PropTypes.array,
  node: PropTypes.object,
};

export default RequestBodyContent;
