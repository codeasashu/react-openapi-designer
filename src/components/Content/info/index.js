import React from 'react';
import PropTypes from 'prop-types';
import {observer} from 'mobx-react-lite';
import AutosizeInput from 'react-input-autosize';
import {InputGroup} from '@blueprintjs/core';
import {MarkdownEditor} from '../../Editor';
import Servers from './servers';
import SecuritySchemes from './security';
import Contact from './contact';
import License from './license';
import {usePatchOperation, getValueFromStore} from '../../../utils/selectors';
import {nodeOperations} from '../../../datasets/tree';

const Info = observer(() => {
  const handlePatch = usePatchOperation();
  return (
    <div className="flex-1 relative">
      <div className="EditorPanel EditorPanel--primary EditorPanel--forms group p-0 flex flex-col relative inset-0">
        <div className="w-10/12 m-auto p-10">
          <div className="flex items-center mb-8">
            <div className="FormHttpService__version bp3-tag bg-green p-2">
              <div className="FormEditableText text-xl font-bold">
                <AutosizeInput
                  inputClassName="bg-transparent hover:bg-darken-1 focus:bg-darken-2"
                  minWidth={30}
                  key={'/'}
                  value={getValueFromStore(['info', 'version']) || '0.1'}
                  onChange={(e) => {
                    handlePatch(
                      nodeOperations.Replace,
                      ['info', 'version'],
                      e.target.value,
                    );
                  }}
                />
              </div>
            </div>
            <input
              className="px-2 py-1 font-medium bg-transparent hover:bg-darken-2 focus:bg-darken-2 rounded-lg ml-4 flex-1 text-2xl"
              value={getValueFromStore(['info', 'title']) || ''}
              placeholder="Name"
              onChange={(e) =>
                handlePatch(
                  nodeOperations.Replace,
                  ['info', 'title'],
                  e.target.value,
                )
              }
            />
          </div>
          <div className="font-semibold text-gray-6 dark:text-lighten-8 ml-1 mt-8 mb-2">
            Description
          </div>
          <div className="flex-1">
            <MarkdownEditor
              value={getValueFromStore(['info', 'description']) || ''}
              placeholder="API description..."
              onChange={(e) =>
                handlePatch(nodeOperations.Replace, ['info', 'description'], e)
              }
            />
          </div>
          <div className="my-8 -mx-1 border-t dark:border-darken-4" />
          <Servers
            servers={getValueFromStore(['servers']) || []}
            onChange={(e) => {
              handlePatch(nodeOperations.Replace, ['servers'], e);
            }}
          />
          <div className="my-8 -mx-1 border-t dark:border-darken-4" />
          <SecuritySchemes
            schemes={getValueFromStore(['components', 'securitySchemes'])}
            onChange={(schemes) => {
              handlePatch(
                nodeOperations.Replace,
                ['components', 'securitySchemes'],
                schemes,
              );
            }}
          />
          <div className="my-8 -mx-1 border-t dark:border-darken-4" />
          <Contact
            contact={getValueFromStore(['info', 'contact'])}
            onChange={(e) => {
              handlePatch(nodeOperations.Replace, ['info', 'contact'], e);
            }}
          />
          <InputGroup
            className="flex-1 mt-4 pr-0"
            title="Terms of Service URL"
            placeholder="Terms of Service URL"
            value={getValueFromStore(['info', 'termsOfService']) || ''}
            onChange={(e) => {
              handlePatch(
                nodeOperations.Replace,
                ['info', 'termsOfService'],
                e.target.value,
              );
            }}
          />
          <div className="my-8 -mx-1 border-t dark:border-darken-4" />
          <License
            license={getValueFromStore(['info', 'license'])}
            onChange={(license) => {
              handlePatch(nodeOperations.Replace, ['info', 'license'], license);
            }}
          />
          <div className="my-8 -mx-1 border-t dark:border-darken-4 mb-5" />
        </div>
      </div>
    </div>
  );
});

Info.propTypes = {
  relativeJsonPath: PropTypes.array,
};

export default Info;
