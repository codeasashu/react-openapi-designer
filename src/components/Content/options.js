import React from 'react';
import PropTypes from 'prop-types';
import {observer} from 'mobx-react-lite';
import classnames from 'classnames';
import {Button, ButtonGroup, Icon, Intent} from '@blueprintjs/core';
import {StoresContext} from '../Context';
import {NodeCategories, NodeTypes} from '../../datasets/tree';
import Tags from './tags';

const Options = observer(({relativeJsonPath, node, onDelete}) => {
  const stores = React.useContext(StoresContext);
  const {activeWidget, activeView, widgets, views} = stores.uiStore;
  const [confirmDelete, setConfirmDelete] = React.useState(false);
  const {errors, hints, info, warning} = stores.lintStore;

  const lintSpec = () => {
    if (activeWidget === widgets.lint) {
      stores.uiStore.toggleWidget(stores.uiStore.widgets.lint);
    } else {
      stores.uiStore.setActiveWidget(stores.uiStore.widgets.lint);
    }
  };

  return (
    <div
      className={classnames(
        'border-b  PanelActionBar w-full flex items-center px-5 py-2',
        {
          'border-transparent': !activeWidget,
        },
      )}>
      <div className="flex items-center">
        <Tags relativeJsonPath={relativeJsonPath} node={node} />
        {node && node.category === NodeCategories.SourceMap && (
          <Button
            small
            icon={<Icon size={14} icon="trash" />}
            text={confirmDelete ? 'Are you sure?' : 'delete'}
            onMouseLeave={() => {
              setConfirmDelete(false);
            }}
            onClick={() => {
              if (confirmDelete) {
                if (onDelete) {
                  onDelete();
                }
              } else {
                setConfirmDelete(true);
              }
            }}
          />
        )}
      </div>
      <div className="flex-1" />
      <div>
        <ButtonGroup>
          <Button
            active={activeView === views.form}
            small
            icon={<Icon size={14} icon="form" />}
            text="Form"
            onClick={() => stores.uiStore.setActiveView(views.form)}
          />
          <Button
            active={activeView === views.code}
            small
            icon={<Icon size={14} icon="code" />}
            text="Code"
            onClick={() => stores.uiStore.setActiveView(views.code)}
          />
        </ButtonGroup>
      </div>
      <div className="ml-3">
        <Button
          active={activeView === views.preview}
          small
          icon={<Icon size={14} icon="document" />}
          text="Preview"
          onClick={() => stores.uiStore.setActiveView(views.preview)}
        />
      </div>
      <div className="ml-3">
        <ButtonGroup>
          {node && node.type === NodeTypes.Operation && (
            <Button
              active={activeWidget === widgets.samples}
              small
              icon={<Icon size={14} icon="code" />}
              text="Samples"
              onClick={() => stores.uiStore.toggleWidget(widgets.samples)}
            />
          )}
          <Button
            active={activeWidget === widgets.lint}
            small
            text={
              <div className="flex justify-center">
                <Icon
                  size={14}
                  icon="error"
                  intent={Intent.DANGER}
                  className="m-auto pl-2 pr-1"
                />
                <span className="counter m-auto">{errors.length}</span>
                <Icon
                  size={14}
                  intent={Intent.WARNING}
                  icon="warning-sign"
                  className="m-auto pl-2 pr-1"
                />
                <span className="counter m-auto">{warning.length}</span>
                <Icon
                  size={14}
                  icon="info-sign"
                  intent={Intent.PRIMARY}
                  className="m-auto pl-2 pr-1"
                />
                <span className="counter m-auto">{info.length}</span>
                <Icon
                  size={14}
                  icon="lightbulb"
                  intent={Intent.SUCCESS}
                  className="m-auto pl-2 pr-1"
                />
                <span className="counter m-auto">{hints.length}</span>
              </div>
            }
            onClick={() => lintSpec()}
          />
        </ButtonGroup>
      </div>
      <div className="ml-3">
        <Button
          small
          aria-label="toggle fullscreen"
          icon={
            <Icon
              size={14}
              icon={stores.uiStore.fullscreen === true ? 'minimize' : 'move'}
            />
          }
          onClick={() => stores.uiStore.toggleFullscreen()}
        />
      </div>
    </div>
  );
});

Options.propTypes = {
  view: PropTypes.string,
  onToggleView: PropTypes.func,
  onDelete: PropTypes.func,
};

export default Options;
