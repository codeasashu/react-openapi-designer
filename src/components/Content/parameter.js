//@flow
import React from 'react';
import PropTypes from 'prop-types';
import {observer} from 'mobx-react-lite';
import {getValueFromStore} from '../../utils/selectors';
import Parameter from '../Common/Parameters/parameter';
import {StoresContext} from '../Context';

const ParameterContent = observer(({relativeJsonPath, node}) => {
  const stores = React.useContext(StoresContext);
  const parameterType = getValueFromStore(relativeJsonPath.concat(['in']));
  return parameterType ? (
    <div className="flex-1 relative">
      <div className="EditorPanel EditorPanel--primary EditorPanel--forms group p-0 flex flex-col relative inset-0">
        <div className="w-full max-w-6xl m-auto p-10 flex flex-col">
          <div className="capitalize px-2 py-1 mb-6 font-medium text-2xl">
            {parameterType} Parameter
          </div>
          <Parameter
            parameterPath={relativeJsonPath}
            parameterIn={parameterType}
            nameInPath={false}
            handleRemove={async () => {
              await stores.graphStore.removeNode(node.id);
            }}
            typePath={relativeJsonPath.concat(['schema', 'type'])}
          />
        </div>
      </div>
    </div>
  ) : null;
});

ParameterContent.propTypes = {
  relativeJsonPath: PropTypes.array,
  node: PropTypes.object,
};

export default ParameterContent;
