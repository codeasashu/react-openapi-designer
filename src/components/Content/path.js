import React, {useState} from 'react';
import {useSelector, useDispatch} from 'react-redux';
import {
  handleModelChange,
  handlePathNameChange,
  handleAddOperation,
} from 'store/modules/openapi';
import PropTypes from 'prop-types';
import {invert, get} from 'lodash';
import {Tab, Tabs, TabList, TabPanel} from 'react-tabs';
import {useHistory, useLocation} from 'react-router-dom';
import {TitleEditor, UrlEditor} from '../Editor';
import {MethodPane} from '../Panes';
import {escapeUri} from '../../utils';

const methodTabClasses =
  'bp3-simple-tab-panel rounded-none flex-1 border-l-0 \
border-r-0 border-b-0 relative';

function useQuery() {
  return new URLSearchParams(useLocation().search);
}

const tabIndexes = {
  get: 0,
  post: 1,
  put: 2,
  delete: 3,
};

const unescapeUri = (path) => path.replaceAll(/~1/g, '/');

const PathContent = ({path, relativeJsonPath, ...props}) => {
  let relativePath = relativeJsonPath.map((i) => unescapeUri(i));
  const dispatch = useDispatch();
  let history = useHistory();

  const onChange = React.useCallback(
    (path, value) => dispatch(handleModelChange({path, value})),
    [dispatch],
  );

  let query = useQuery();
  const pathItem = useSelector(({openapi}) =>
    get(openapi, relativePath.slice(0, 2)),
  );

  const selectDefaultMethod = () => {
    const methods = Object.keys(pathItem).map((m) => m.toLowerCase());
    if (methods.length === 0) return null;
    if (methods.indexOf('get') > -1) return 'get';
    if (methods.indexOf('post') > -1) return 'post';
    if (methods.indexOf('put') > -1) return 'put';
    if (methods.indexOf('delete') > -1) return 'delete';
    return methods[0].toLowerCase();
  };

  const addOperation = (method) => {
    dispatch(handleAddOperation({path: relativePath[1], method}));
    query.set(
      'path',
      `paths/${escapeUri(relativePath[1])}/${method.toLowerCase()}`,
    );
    history.push(`/designer?${query}`);
  };

  const updateOperation = (methodName, operation) => {
    onChange({
      path,
      pathItem: {
        ...pathItem,
        [methodName.toLowerCase()]: {...operation, ...operation},
      },
    });
  };

  const onTabClick = (selectedIndex) => {
    setSelectedTab(selectedIndex);
    const methodName = invert(tabIndexes)[selectedIndex];
    if (Object.keys(pathItem).indexOf(methodName.toLowerCase()) >= 0) {
      query.set(
        'path',
        `paths/${escapeUri(relativePath[1])}/${methodName.toLowerCase()}`,
      );
      history.push(`/designer?${query}`);
    }
  };

  const extractMethodFromUri = (validMethods, relativePath) => {
    if (relativePath.length < 3) return null;
    const method = relativePath[2];
    const isValidMethod = validMethods.indexOf(method.toLowerCase()) >= 0;
    return isValidMethod ? method.toLowerCase() : null;
  };

  let method = extractMethodFromUri(Object.keys(tabIndexes), relativePath);
  if (method === null) {
    method = selectDefaultMethod();
    const copyOfRelativePath = relativePath.slice(0, 2);
    copyOfRelativePath.push(method);
    relativePath = copyOfRelativePath;
  }

  const operation = useSelector(({openapi}) => {
    return get(openapi, relativePath);
  });

  const [selectedTab, setSelectedTab] = useState(
    tabIndexes[method.toLowerCase()],
  );

  const onPathChange = React.useCallback(
    (newPath, oldPath) => {
      dispatch(handlePathNameChange({newPath, oldPath}));
      const methodName = invert(tabIndexes)[selectedTab];
      query.set(
        'path',
        `#/paths/${escapeUri(newPath)}/${methodName.toLowerCase()}`,
      );
      history.replace(`/designer?${query}`);
    },
    [dispatch],
  );

  return (
    <div className="flex-1 relative">
      <div className="EditorPanel EditorPanel--primary EditorPanel--forms group p-0 flex flex-col relative inset-0">
        <div className="flex px-10 mt-10 max-w-6xl justify-between">
          <TitleEditor
            xl
            value={operation['summary']}
            placeholder="Operation Name"
            onChange={(e) => {
              const summary = e.target.value;
              onChange(relativePath.concat(['summary']), summary);
            }}
          />
        </div>
        <div className="px-10 pb-2 max-w-6xl">
          <div className="mt-6">
            <UrlEditor
              errors={props.errors?.url}
              value={relativePath[1]}
              onChange={(newPath) => {
                onPathChange(newPath, relativePath[1]);
              }}
            />
          </div>
        </div>
        <Tabs
          className="flex flex-col flex-1"
          selectedTabClassName="selected-tab"
          selectedTabPanelClassName="block"
          selectedIndex={selectedTab}
          onSelect={onTabClick}>
          <TabList className="mt-6 px-10 flex bp3-simple-tab-list">
            <Tab
              className={`bp3-simple-tab uppercase ${
                method.toLowerCase() === 'get' && 'text-green-400'
              }`}>
              get
            </Tab>
            <Tab
              className={`bp3-simple-tab uppercase ${
                method.toLowerCase() === 'post' && 'text-blue-600'
              }`}>
              post
            </Tab>
            <Tab
              className={`bp3-simple-tab uppercase ${
                method.toLowerCase() === 'put' && 'text-yellow-600'
              }`}>
              put
            </Tab>
            <Tab
              className={`bp3-simple-tab uppercase ${
                method.toLowerCase() === 'delete' && 'text-red-400'
              }`}>
              delete
            </Tab>
          </TabList>
          <TabPanel className={methodTabClasses}>
            <MethodPane
              methodName="get"
              operation={pathItem['get']}
              onAddOperation={() => addOperation('get')}
              onChange={(e) => updateOperation('get', e)}
            />
          </TabPanel>
          <TabPanel className={methodTabClasses}>
            <MethodPane
              methodName="post"
              operation={pathItem['post']}
              onAddOperation={() => addOperation('post')}
              onChange={(e) => updateOperation('post', e)}
            />
          </TabPanel>
          <TabPanel className={methodTabClasses}>
            <MethodPane
              methodName="put"
              operation={pathItem['put']}
              onAddOperation={() => addOperation('put')}
              onChange={(e) => updateOperation('put', e)}
            />
          </TabPanel>
          <TabPanel className={methodTabClasses}>
            <MethodPane
              methodName="delete"
              operation={pathItem['delete']}
              onAddOperation={() => addOperation('delete')}
              onChange={(e) => updateOperation('delete', e)}
            />
          </TabPanel>
        </Tabs>
      </div>
    </div>
  );
};

PathContent.propTypes = {
  path: PropTypes.string,
  method: PropTypes.string,
  relativeJsonPath: PropTypes.array,
  errors: PropTypes.any,
  onChange: PropTypes.func,
};

export default PathContent;
