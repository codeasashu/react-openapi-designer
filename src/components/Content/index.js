import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {observer} from 'mobx-react-lite';
import {cloneDeep} from 'lodash';
import Redoc from '../Docs';
import Options from './options';
import Info from './info';
import PathContent from './path';
import ModelContent from './model';
import Parameter from './parameter';
import Response from './response';
import RequestBody from './RequestBody';
import MonacoEditor from '../Editor/Monaco';
import LintWidget from '../Widgets/Lint';
import SamplesWidget from '../Widgets/Samples';
import {StoresContext} from '../Context';
import {NodeCategories, NodeTypes} from '../../datasets/tree';

const StyledContent = styled.div`
  width: calc(80% - 2px);
`;

const getComponentForNode = (node) => {
  switch (node.type) {
    case NodeTypes.Path:
    case NodeTypes.Operation:
      return PathContent;
    case NodeTypes.Response:
      return Response;
    case NodeTypes.RequestBody:
      return RequestBody;
    case NodeTypes.Model:
      return ModelContent;
    case NodeTypes.Parameter:
      return Parameter;
    case NodeTypes.Overview:
    default:
      return Info;
  }
};

const SubContent = observer(({node, relativeJsonPath}) => {
  const stores = React.useContext(StoresContext);
  const {activeWidget, widgets} = stores.uiStore;
  const RenderSubContent = getComponentForNode(node);
  return activeWidget ? (
    <div className="flex-1 relative flex">
      <RenderSubContent relativeJsonPath={relativeJsonPath} node={node} />
      <div className="relative flex-1 border-l">
        {activeWidget === widgets.lint && <LintWidget />}
        {activeWidget === widgets.samples && <SamplesWidget />}
      </div>
    </div>
  ) : (
    <RenderSubContent relativeJsonPath={relativeJsonPath} node={node} />
  );
});

SubContent.propTypes = {
  node: PropTypes.object,
};

const Content = observer(() => {
  const stores = React.useContext(StoresContext);
  const {activeNode, activeView, views} = stores.uiStore;
  const sourceNode = stores.graphStore.rootNode;
  const node = activeNode || sourceNode;
  const relativeJsonPath =
    node.category === NodeCategories.SourceMap ? node.relativeJsonPath : [];

  return (
    <StyledContent className={'flex flex-col flex-1'}>
      <div className="bp3-dark relative flex flex-1 flex-col bg-canvas">
        <Options
          relativeJsonPath={relativeJsonPath}
          node={node}
          onDelete={() => {
            stores.graphStore.removeNode(activeNode.id);
          }}
        />
        {activeView === views.form && (
          <SubContent node={node} relativeJsonPath={relativeJsonPath} />
        )}
        {activeView === views.code && <MonacoEditor />}
        {activeView === views.preview && (
          <Redoc spec={cloneDeep(sourceNode.data.parsed)} />
        )}
      </div>
    </StyledContent>
  );
});

Content.propTypes = {};

export default Content;
