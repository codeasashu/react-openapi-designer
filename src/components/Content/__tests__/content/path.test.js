import React from 'react';
import userEvent from '@testing-library/user-event';
import {fireEvent, prettyDOM} from '@testing-library/dom';
import {
  act,
  render,
  screen,
  StoreCreator,
  within,
} from '../../../../../test-utils';
import Content from '../../index';

describe('Path content test', () => {
  const getTitleNode = () => screen.getByPlaceholderText(/Operation Name/);
  const getMethodTabs = () => screen.getAllByRole(/tab/);
  const getPathNameNode = () => screen.getByLabelText(/path/);
  const getOperationNode = (method) => screen.getByRole(`operation-${method}`);
  const getParameters = (method, paramType) => {
    return (
      within(getOperationNode(method)).queryAllByLabelText(paramType) || []
    );
  };

  const getParamInput = (method, paramType) => {
    const inpContainer = within(
      within(getOperationNode(method)).getByLabelText(paramType),
    );
    return {
      name: inpContainer.getByPlaceholderText(/Name/),
    };
  };

  const getRequestBodyNode = () => screen.getByLabelText(/request-body/);
  const getResponseStatusNode = () =>
    screen.getByLabelText(/response-statuscode/);

  const assertTitle = (expected) => {
    expect(getTitleNode()).toHaveValue(expected);
  };

  const assertMethodName = (expected) => {
    const methodTabNodes = getMethodTabs();
    const activeMethodNode = methodTabNodes.find((tabNode) => {
      return tabNode.classList.contains('selected-tab');
    });
    expect(activeMethodNode).toHaveTextContent(expected);
  };

  const assertPathName = (expected) => {
    expect(getPathNameNode()).toHaveValue(expected);
  };

  const assertOperationId = (method, expected) => {
    expect(
      within(getOperationNode(method)).getByRole(/operation-id/),
    ).toHaveTextContent(expected);
  };

  const assertDescription = (method, expected) => {
    expect(
      within(getOperationNode(method)).getByPlaceholderText(
        /Endpoint description.../,
      ),
    ).toHaveTextContent(expected);
  };

  const assertParameterLength = (method, paramType, expected) => {
    const params = getParameters(method, paramType);
    expect(params).toHaveLength(expected);
  };

  const assertBody = (method) => {
    const requestBodyNode = getRequestBodyNode();
    const descriptionNode =
      within(requestBodyNode).getByPlaceholderText(/description/);
    const schemaRows = within(requestBodyNode).getAllByRole(/schema-row/);
    return {
      description: (expected) => {
        expect(descriptionNode).toHaveValue(expected);
      },
      schemaRowLength: (expected) => {
        expect(schemaRows).toHaveLength(expected);
      },
    };
  };

  const assertResponseStatusCode = (codes = []) => {
    expect(within(getResponseStatusNode()).getAllByRole(/button/)).toHaveLength(
      codes.length + 1, // 1 for "Add response button" + length of existing status code
    );
  };

  const assertParameter = async (method, paramType, targetJson) => {};

  it('tests default created schema', async () => {
    const {stores, creator, asserts} = StoreCreator();
    const path = creator.createPath('/user/abc');
    stores.uiStore.setActiveNode(path);
    const {container, asFragment} = render(<Content />, {
      providerProps: {value: stores},
    });

    expect(asFragment).toMatchSnapshot();
    assertTitle('Your GET endpoint');
    assertMethodName('get');
    assertPathName('/user/abc');
    assertOperationId('get', 'get-user-abc');
    assertDescription('get', '');
    assertParameterLength('get', 'header', 0);
    assertParameterLength('get', 'query', 0);
    assertParameterLength('get', 'security', 0);
    assertParameterLength('get', 'cookie', 0);
    assertResponseStatusCode([]);
    expect(screen.queryByPlaceholderText(/Response description/)).toBeNull();
    const doc = asserts.oas(path.relativeJsonPath, true);
    expect(doc).toStrictEqual({
      operationId: 'get-user-abc',
      responses: {},
      summary: 'Your GET endpoint',
      tags: [],
    });
  });

  it('can change summary of path', async () => {
    const {stores, creator, asserts} = StoreCreator();
    const path = creator.createPath('/user/abc');
    stores.uiStore.setActiveNode(path);
    const {container} = render(<Content />, {providerProps: {value: stores}});
    assertTitle('Your GET endpoint');
    await act(async () => {
      await fireEvent.change(getTitleNode(), {
        target: {value: 'Abcd'},
      });
    });
    assertTitle('Abcd');
    const doc = asserts.oas(path.relativeJsonPath, true);
    expect(doc).toStrictEqual({
      operationId: 'get-user-abc',
      responses: {},
      summary: 'Abcd',
      tags: [],
    });
  });

  it('can add parameters to path', async () => {
    const {stores, creator, asserts} = StoreCreator();
    const path = creator.createPath('/user/abc');
    stores.uiStore.setActiveNode(path);
    const {container} = render(<Content />, {providerProps: {value: stores}});
    act(() => {
      fireEvent.click(screen.getByRole(/header/, {text: 'Header'}));
    });

    assertParameterLength('get', 'header', 1);
    assertParameterLength('get', 'query', 0);
    assertParameterLength('get', 'security', 0);
    assertParameterLength('get', 'cookie', 0);

    expect(asserts.oas(path.relativeJsonPath, true)).toStrictEqual({
      operationId: 'get-user-abc',
      parameters: [
        {
          in: 'header',
          schema: {
            type: 'string',
          },
        },
      ],
      responses: {},
      summary: 'Your GET endpoint',
      tags: [],
    });

    // Assert Header param
    await act(async () => {
      await fireEvent.change(getParamInput('get', 'header').name, {
        target: {value: 'Abcd'},
      });
      fireEvent.focusOut(getParamInput('get', 'header').name);
    });
    expect(
      asserts.oas(path.relativeJsonPath.concat('parameters'), true),
    ).toContainObject({
      in: 'header',
      name: 'Abcd',
      schema: {
        type: 'string',
      },
    });
    assertParameterLength('get', 'header', 1);

    //Add query parameters
    act(() => {
      fireEvent.click(screen.getByRole(/query/, {text: /Query/}));
    });

    await act(async () => {
      await fireEvent.change(getParamInput('get', 'query').name, {
        target: {value: 'ppp'},
      });
      fireEvent.focusOut(getParamInput('get', 'query').name);
    });

    expect(
      asserts.oas(path.relativeJsonPath.concat('parameters'), true),
    ).toContainObject({
      in: 'query',
      name: 'ppp',
      schema: {
        type: 'string',
      },
    });
    assertParameterLength('get', 'header', 1);
  });

  it('can add request body', async () => {
    const {stores, creator, asserts} = StoreCreator();
    const path = creator.createPath('/user/abc');
    stores.uiStore.setActiveNode(path);
    render(<Content />, {providerProps: {value: stores}});
    const requestBodyNode = getRequestBodyNode();
    const addBtn = within(requestBodyNode).getByRole(/button/, {
      name: /Add Body/,
    });
    act(() => {
      fireEvent.click(addBtn);
    });
    expect(
      within(requestBodyNode).getByPlaceholderText(/description/),
    ).toHaveValue('');
    const schemaRows = within(requestBodyNode).getAllByRole(/schema-row/);
    expect(schemaRows).toHaveLength(1);
    expect(schemaRows[0]).toHaveTextContent(/object/);
    expect(
      within(requestBodyNode).getByPlaceholderText(/Create or choose existing/),
    ).toHaveValue('application/json');
    expect(within(requestBodyNode).getByRole(/combobox/)).toHaveValue(
      'application/json',
    );
    expect(
      asserts.oas(path.relativeJsonPath.concat('requestBody'), true),
    ).toStrictEqual({
      content: {
        'application/json': {
          schema: {
            type: 'object',
            properties: {},
          },
        },
      },
    });
  });

  describe('Responses', () => {
    it('does not have responses node by default', () => {
      const {stores, creator, asserts} = StoreCreator();
      const path = creator.createPath('/user/abc');
      stores.uiStore.setActiveNode(path);
      render(<Content />, {providerProps: {value: stores}});
      const responseStatusNode = getResponseStatusNode();
      const responsesNode = screen.getByRole(/responses/);
      expect(responsesNode.querySelectorAll('.response').length).toBe(0);
      expect(
        within(responseStatusNode).queryByRole(/button/, {
          name: /200/,
        }),
      ).toBeNull();
    });

    it('can add responses', async () => {
      const {stores, creator, asserts} = StoreCreator();
      const path = creator.createPath('/user/abc');
      stores.uiStore.setActiveNode(path);
      render(<Content />, {providerProps: {value: stores}});
      const responseStatusNode = getResponseStatusNode();
      const responsesNode = screen.getByRole(/responses/);
      const addBtn = within(responseStatusNode).getByRole(/button/, {
        name: /Response/,
      });
      act(() => {
        fireEvent.click(addBtn);
      });
      expect(
        Object.keys(
          asserts.oas(path.relativeJsonPath.concat('responses'), true),
        ),
      ).toStrictEqual(['200']);

      expect(
        within(responsesNode).getByPlaceholderText(/description/),
      ).toHaveValue('');

      expect(
        within(responseStatusNode).getByRole(/button/, {
          name: /200/,
        }),
      ).toBeInTheDocument();

      const addBodyBtn = within(responsesNode).getByRole(/button/, {
        name: /Add Body/,
      });
      act(() => {
        fireEvent.click(addBodyBtn);
      });

      const schemaRows = within(responsesNode).getAllByRole(/schema-row/);
      expect(schemaRows).toHaveLength(1);
      expect(schemaRows[0]).toHaveTextContent(/object/);
      expect(
        within(responsesNode).getByPlaceholderText(/Create or choose existing/),
      ).toHaveValue('application/json');
      expect(within(responsesNode).getByRole(/combobox/)).toHaveValue(
        'application/json',
      );
    });

    it('replace selected content type on selecting', async () => {
      const {stores, creator, asserts} = StoreCreator();
      const path = creator.createPath('/user/abc');
      stores.uiStore.setActiveNode(path);
      const {asFragment, getByRole} = render(<Content />, {
        providerProps: {value: stores},
      });
      const responseStatusNode = getResponseStatusNode();
      const responsesNode = screen.getByRole(/responses/);
      const addBtn = within(responseStatusNode).getByRole(/button/, {
        name: /Response/,
      });
      act(() => {
        fireEvent.click(addBtn);
      });
      const addBodyBtn = within(responsesNode).getByRole(/button/, {
        name: /Add Body/,
      });
      act(() => {
        fireEvent.click(addBodyBtn);
      });

      await act(async () => {
        await userEvent.click(
          screen.getByPlaceholderText(/Create or choose existing/),
        );
        const popover = document.querySelector('.bp3-popover.bp3-minimal');
        expect(popover).not.toBeNull();
        await userEvent.click(
          within(popover).getByTestId('application/xml'),
          undefined,
          {
            skipPointerEventsCheck: true,
          },
        );
      });
      expect(
        screen.getByPlaceholderText(/Create or choose existing/),
      ).toHaveValue('application/xml');
      expect(getByRole(/combobox/)).toHaveValue('application/xml');
      expect(getByRole(/combobox/).getElementsByTagName('option')).toHaveLength(
        1,
      );
      const doc = asserts.oas(
        path.relativeJsonPath.concat(['responses', '200', 'content']),
        true,
      );
      expect(doc).toHaveProperty('application/xml');
      expect(Object.keys(doc)).toHaveLength(1);
      expect(asFragment).toMatchSnapshot();
    });
  });
});
