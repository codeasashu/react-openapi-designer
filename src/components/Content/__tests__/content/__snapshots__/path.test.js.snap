// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Path content test Responses replace selected content type on selecting 1`] = `[Function]`;

exports[`Path content test tests default created schema 1`] = `[Function]`;
